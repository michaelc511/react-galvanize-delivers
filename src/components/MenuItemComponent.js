import React from 'react';

/*
// if not defined, show NA
*/

export default function MenuItemComponent({ item, onAddItem }) {
  if (!item) return null;

  /*
   Local defined function
   1. create handleClick function to hande the event
   2. on the <a> add 'onClick = {<function>}
    */
  function handleClick(event) {
    event.preventDefault();
    //console.log('item: ' + item.id);
    //console.log(event);
    // use the prop fuction to use the item ID
    onAddItem(item.id);
  }

  return (
    <div className="card MenuItem MenuItemComponent">
      <div className="card-image">
        <img src={item.imagePath} alt="Nothing" />
      </div>
      <div className="card-content">
        <h5 className="card-title">
          {item.name || 'N/A'}
        </h5>
        <p>
          {item.price && typeof item.price === 'number' ? `$${item.price.toFixed(2)}` : 'N/A'}
        </p>
      </div>
      <div className="card-action">
        <a href="." onClick={handleClick}>
          ADD TO ORDER
        </a>
      </div>
    </div>
  );
}
