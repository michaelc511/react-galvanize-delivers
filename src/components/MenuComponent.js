import React from 'react';
import MenuItemComponent from './MenuItemComponent';

import './MenuComponent.story.css';

export default function MenuComponent({ items, onAddItem }) {
  //
  if (items && Array.isArray(items)) {
    return (
      <div className="menuItems MenuComponent">
        {items.map((
          item //
        ) =>
          <MenuItemComponent
            key={item.id} //
            item={item} //
            onAddItem={onAddItem}
          />
        )}
      </div>
    );
  } else {
    return <div>Loading...the stuff</div>;
  }
}
