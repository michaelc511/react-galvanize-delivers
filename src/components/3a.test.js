import React from 'react';
import { shallow, mount } from 'enzyme';
import MenuItemComponent from './MenuItemComponent';
import MenuComponent from './MenuComponent';

/*
Phase 3-A

MenuItemComponent Tests
The item's name gets rendered
The item's price gets rendered
The onAddItem callback gets triggered when you click on the "Add to Order" button/link
Note: Shallow rendering should be sufficient
Also, write a snapshot test for MenuItemComponent using Jest.

MenuComponent tests
Verify the expected number of MenuItemComponents are rendered based
on the number of items that are passed into MenuComponent

The onAddItem callback gets triggered when you click on one of the "Add to Order" buttons
*/

const onAddItem = itemId => {
  console.log(`This is the itemID!!! ${itemId}`);
};

const item = { name: 'Some Item A', price: 9.99, imagePath: '//via.placeholder.com/300x200' };

const menuItems = [
  {
    name: 'Smoked Swine',
    price: 8.99,
    imagePath: '//via.placeholder.com/300x200',
    id: 1
  },
  {
    name: 'Royale with Cheese',
    price: 11.99,
    imagePath: '//via.placeholder.com/300x200',
    id: 2
  },
  {
    name: 'Arugula Pie',
    price: 14.99,
    imagePath: '//via.placeholder.com/300x200',
    id: 3
  },
  {
    name: 'Ice Cream Biscuit',
    price: 7.99,
    imagePath: '//via.placeholder.com/300x200',
    id: 4
  }
];

// // 3A Shallow test for name and price are rendered
// describe('a shallow test', () => {
//   const shallowWrapper = shallow(<MenuItemComponent onAddItem={onAddItem} item={item} />);
//
//   it('should have item name and price gets rendered  ', () => {
//     console.log('test');
//     console.log(shallowWrapper);
//     expect(
//       shallowWrapper.contains(
//         <h5>
//           {item.name || 'N/A'}
//         </h5>
//       )
//     ).toBe(true);
//
//     //expect(shallowWrapper.find('item')).toHaveLength(1);
//   });
//
//   it('should have item   price gets rendered  ', () => {
//     console.log('test');
//     console.log(shallowWrapper);
//     expect(
//       shallowWrapper.contains(
//         <p>
//           {item.price || 'N/A'}
//         </p>
//       )
//     ).toBe(true);
//
//     //expect(shallowWrapper.find('item')).toHaveLength(1);
//   });
// });

// 3A render tests using 'mount'
describe('full render tests', () => {
  const fullWrapper = mount(<MenuItemComponent onAddItem={onAddItem} item={item} />);
  it('should test the props of the component', () => {
    expect(fullWrapper.props().item).toEqual(item);
  });
  it("should test the item' name gets rendered ", () => {
    expect(fullWrapper.props().item.name).toEqual(item.name);
  });

  it("should test the item' price gets rendered ", () => {
    expect(fullWrapper.props().item.price).toEqual(item.price);
  });
});

// 3A onAddItem using 'shallow'
describe('a shallow test on event', () => {
  const shallowWrapper = shallow(<MenuItemComponent onAddItem={onAddItem} item={item} />);

  // the onAddItem callback gets triggered when you click on the "Add to Order" button/link
  it('should test if onAddItem function was fired', () => {
    const onAddItem = jest.fn();

    mount(<MenuItemComponent onAddItem={onAddItem} item={item} />).find('.menuChoice').simulate('click');
    expect(onAddItem).toHaveBeenCalled();
  });

  expect(shallowWrapper).toMatchSnapshot();
});

// 3A
//Verify the expected number of MenuItemComponents are rendered based on the number of items that are passed into MenuComponent
//The onAddItem callback gets triggered when you click on one of the "Add to Order" buttons
describe(' tests for MenuComponent', () => {
  it('tests children elemenets with full rendering', () => {
    expect(mount(<MenuComponent items={menuItems} />).children()).toHaveLength(4);
  });

  // the onAddItem callback gets triggered when you click on the "Add to Order" button/link
  it('should test if onAddItem function was fired', () => {
    const onAddItem = jest.fn();

    mount(<MenuItemComponent onAddItem={onAddItem} item={item} />).find('.menuChoice').simulate('click');
    expect(onAddItem).toHaveBeenCalled();
  });
});
