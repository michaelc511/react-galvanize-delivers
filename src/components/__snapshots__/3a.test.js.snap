// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="card MenuItem MenuItemComponent"
>
    <div
        className="card-image"
    >
        <img
            alt="Nothing"
            src="//via.placeholder.com/300x200"
        />
    </div>
    <div
        className="card-content"
    >
        <h5
            className="card-title"
        >
            Some Item A
        </h5>
        <p>
            $9.99
        </p>
    </div>
    <div
        className="card-action"
    >
        <button>
            Button test
        </button>
        <a
            className="menuChoice #fdd835 yellow darken-1 btn"
            href="."
            onClick={[Function]}
        >
            ADD TO ORDER
        </a>
    </div>
</div>,
  "nodes": Array [
    <div
      className="card MenuItem MenuItemComponent"
>
      <div
            className="card-image"
      >
            <img
                  alt="Nothing"
                  src="//via.placeholder.com/300x200"
            />
      </div>
      <div
            className="card-content"
      >
            <h5
                  className="card-title"
            >
                  Some Item A
            </h5>
            <p>
                  $9.99
            </p>
      </div>
      <div
            className="card-action"
      >
            <button>
                  Button test
            </button>
            <a
                  className="menuChoice #fdd835 yellow darken-1 btn"
                  href="."
                  onClick={[Function]}
            >
                  ADD TO ORDER
            </a>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <MenuItemComponent
        item={
                Object {
                        "imagePath": "//via.placeholder.com/300x200",
                        "name": "Some Item A",
                        "price": 9.99,
                      }
        }
        onAddItem={[Function]}
/>,
      "_debugID": 14,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "item": Object {
            "imagePath": "//via.placeholder.com/300x200",
            "name": "Some Item A",
            "price": 9.99,
          },
          "onAddItem": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="card MenuItem MenuItemComponent"
>
          <div
                    className="card-image"
          >
                    <img
                              alt="Nothing"
                              src="//via.placeholder.com/300x200"
                    />
          </div>
          <div
                    className="card-content"
          >
                    <h5
                              className="card-title"
                    >
                              Some Item A
                    </h5>
                    <p>
                              $9.99
                    </p>
          </div>
          <div
                    className="card-action"
          >
                    <button>
                              Button test
                    </button>
                    <a
                              className="menuChoice #fdd835 yellow darken-1 btn"
                              href="."
                              onClick={[Function]}
                    >
                              ADD TO ORDER
                    </a>
          </div>
</div>,
        "_debugID": 15,
        "_renderedOutput": <div
          className="card MenuItem MenuItemComponent"
>
          <div
                    className="card-image"
          >
                    <img
                              alt="Nothing"
                              src="//via.placeholder.com/300x200"
                    />
          </div>
          <div
                    className="card-content"
          >
                    <h5
                              className="card-title"
                    >
                              Some Item A
                    </h5>
                    <p>
                              $9.99
                    </p>
          </div>
          <div
                    className="card-action"
          >
                    <button>
                              Button test
                    </button>
                    <a
                              className="menuChoice #fdd835 yellow darken-1 btn"
                              href="."
                              onClick={[Function]}
                    >
                              ADD TO ORDER
                    </a>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <MenuItemComponent
    item={
        Object {
            "imagePath": "//via.placeholder.com/300x200",
            "name": "Some Item A",
            "price": 9.99,
          }
    }
    onAddItem={[Function]}
/>,
}
`;
